src/services/blocks/trace/Trace.ts

```diff
@@ -0,0 +1,690 @@
+			throw new InternalServerError(
+				'No spans found. This runtime is likely not supported with tracing.'
+			);
+		}
+
+		this.keyNames = getKeyNames(api);
+	}
+
+	/**
+	 * Derive the action groups and operations from the block trace.
+	 *
+	 * @returns the action groups and opertions
+	 */
+	actionsAndOps(): { actions: ActionGroup[]; operations: Operation[] } {
+		const { spansById, parsedSpans: spans, executeBlockSpanId } = parseSpans(
+			this.traceBlock.spans
+		);
+		const { eventsByParentId, extrinsicIndexBySpanId } = this.parseEvents(
+			spansById
+		);
+
+		const operations: Operation[] = [];
+		const actions: ActionGroup[] = [];
+		// Create a list of action groups (`actions`) and a list of all `Operations`.
+		for (const primary of spans) {
+			if (!(primary.parentId === executeBlockSpanId)) {
+				// Only use primary spans (spans where parentId === execute_block). All other
+				// spans are either the executeBlockSpan or descendant of a primary span.
+				continue;
+			}
+
+			// Gather the secondary spans and their events. Events are useful for
+			// creating Operations. Spans are useful for attributing events to a
+			// hook or extrinsic.
+			// Keep in mind events are storage Get/Put events. The spans are used to
+			// attribute events to some action in the block execution pipeline.
+			const secondarySpansIds = findDescendants(primary.id, spansById);
+			const secondarySpansEvents: PEvent[] = [];
+			const secondarySpans: SpanWithChildren[] = [];
+			secondarySpansIds.forEach((id) => {
+				const events = eventsByParentId.get(id);
+				const span = spansById.get(id);
+
+				events && secondarySpansEvents.push(...events);
+				span && secondarySpans.push(span);
+			});
+
+			let phase: PhaseOther, extrinsicIndex: BN | undefined;
+			if (primary.name === SPANS.applyExtrinsic.name) {
+				// This is an action group for an extrinsic application
+				extrinsicIndex = extractExtrinsicIndex(
+					// Pass in all the relevant spanIds of this action group
+					secondarySpansIds.concat(primary.id),
+					extrinsicIndexBySpanId
+				);
+
+				phase = Phase.ApplyExtrinsic;
+			} else {
+				// This is likely an action group for a pallet hook or some initial/final
+				// block execution checks (i.e one of `initBlock`, `initialChecks`,
+				// `onFinalize`, or `finalChecks`).
+				// Note: `onInitialize` spans belong to the `initBlock` action group, so
+				// we give that action group the phase `onInitialize`.
+
+				phase = secondarySpans[0]
+					? // This case catches `onInitialize` spans since they are the secondary
+					  // spans of `initBlock`
+					  stringCamelCase(secondarySpans[0]?.name)
+					: // This case catches `onFinalize` since `onFinalize` spans are
+					  // identified by the priamry spans name.
+					  stringCamelCase(primary.name);
+			}
+
+			const events =
+				eventsByParentId
+					.get(primary.id)
+					?.concat(secondarySpansEvents)
+					.sort((a, b) => a.eventIndex - b.eventIndex) || [];
```

Comment:
Can probably remove this sort and the sort when parsing spans. Instead we can just sort all the Operations at the end by eventIndex before returning. Although it may be less stable for in terms of Operation index in the case when we get multiple Operations for an event

Reply:
Removing this sort breaks things even if we sort events for each address in `accountEventsByAddress`, not sure why.